src/libsync/clientsideencryption.cpp

```diff
@@ -0,0 +1,1168 @@
+            handleErrors();
+        }
+
+        /* Initialise the decryption operation. */
+        if(!EVP_DecryptInit_ex(ctx, EVP_aes_128_gcm(), NULL, NULL, NULL)) {
+			qCInfo(lcCse()) << "Error initializing the decryption context";
+            handleErrors();
+        }
+
+        /* Set IV length to 16 bytes */
+        if(!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, 16, NULL)) {
+			qCInfo(lcCse()) << "Error seting th iv length for the decrypt context";
+            handleErrors();
+        }
+
+        /* Initialise key and IV */
+        if(!EVP_DecryptInit_ex(ctx, NULL, NULL, key, iv)) {
+			qCInfo(lcCse()) << "Error setting the key and iv for decryption";
+            handleErrors();
+        }
+
+       /* Provide the message to be decrypted, and obtain the plaintext output.
+        * EVP_DecryptUpdate can be called multiple times if necessary
+        *
+        * Do not try to decrypt the last 16 bytes. The tag is appended by Android.
+        * So we ignore the last 16 bytes.
+        */
+        if(!EVP_DecryptUpdate(ctx, plaintext, &len, ciphertext, ciphertext_len - 16)) {
+			qCInfo(lcCse()) << "Error decrypting the text";
+            handleErrors();
+        }
+        plaintext_len = len;
+
+        /* Set expected tag value. Works in OpenSSL 1.0.1d and later */
+        if(!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_TAG, 16, tag)) {
+			qCInfo(lcCse()) << "Error setting the tag on the decrupt context";
+            handleErrors();
+        }
+
+        /* Finalise the decryption. A positive return value indicates success,
+        * anything else is a failure - the plaintext is not trustworthy.
+        */
+        ret = EVP_DecryptFinal_ex(ctx, plaintext + len, &len);
+
+        /* Clean up */
+        EVP_CIPHER_CTX_free(ctx);
+
+        if(ret > 0)
+        {
+            /* Success */
+            plaintext_len += len;
+            return plaintext_len;
+        }
+        else
+        {
+			qCInfo(lcCse()) << "Error finalizing the decrypt";
+            /* Verify failed */
+            return -1;
+        }
+    }
+}
+
+ClientSideEncryption::ClientSideEncryption()
+{
+}
+
+void ClientSideEncryption::setAccount(AccountPtr account)
+{
+    _account = account;
+}
+
+void ClientSideEncryption::initialize()
+{
+    qCInfo(lcCse()) << "Initializing";
+    if (!_account->capabilities().clientSideEncryptionAvaliable()) {
+        qCInfo(lcCse()) << "No Client side encryption avaliable on server.";
+        emit initializationFinished();
+    }
+
+    if (hasPrivateKey() && hasPublicKey()) {
+        qCInfo(lcCse()) << "Public and private keys already downloaded";
+        emit initializationFinished();
+    }
+
+    getPublicKeyFromServer();
+}
+
+QString publicKeyPath(AccountPtr account)
+{
+    return baseDirectory() + account->displayName() + ".pub";
+}
+
+QString privateKeyPath(AccountPtr account)
+{
+    return baseDirectory() + account->displayName() + ".rsa";
```

Comment:
Instead of storing the key pairs unprotected on the filesystem we should use https://github.com/frankosterfeld/qtkeychain